app/javascript/lib/imperial.js

```diff
@@ -1834,91 +1869,60 @@ export default class Imperial {
-        rondelPositions[currentIndex + 3] || rondelPositions[distance + 3],
-      ].forEach((slot) => {
-        out.add(Action.rondel({ nation, cost: 0, slot }));
-      });
-      out.add(
-        Action.rondel({
-          nation,
-          cost: this.extraRondelCost(1),
-          slot:
-            rondelPositions[currentIndex + 4] || rondelPositions[distance + 4],
-        }),
-      );
-      out.add(
-        Action.rondel({
-          nation,
-          cost: this.extraRondelCost(2),
-          slot:
-            rondelPositions[currentIndex + 5] || rondelPositions[distance + 5],
-        }),
-      );
-      out.add(
+  availableRondelActions(nationName) {
+    const nation = this.nations.get(nationName);
+    const costPerPaidDistance = this.costPerPaidRondelAction(nation);
+
+    const nationCurrentRondelSlot = this.rondel.representationToEntity(nation.rondelPosition);
+    const availableSlots = new AvailableSlots(this.rondel, 3, 3, costPerPaidDistance);
+
+    const availableRondelSlots = new Set();
+
+    const nextAvailableFreeRondelSlots = nationCurrentRondelSlot
+      ? availableSlots.nextAvailableFreeRondelSlots(nationCurrentRondelSlot) : this.rondel.slotOrder;
+    for (const freeRondelSlot of nextAvailableFreeRondelSlots) {
+      availableRondelSlots.add(
         Action.rondel({
-          nation,
-          cost: this.extraRondelCost(3),
-          slot:
-            rondelPositions[currentIndex + 6] || rondelPositions[distance + 6],
+          nation: nationName,
+          cost: 0,
+          slot: freeRondelSlot.representation,
         }),
       );
-    } else {
-      rondelPositions.forEach((slot) => {
-        out.add(Action.rondel({ nation, cost: 0, slot }));
-      });
     }
+
+    if (nationCurrentRondelSlot) {
+      for (const [paidRondelSlot, cost] of availableSlots.nextAvailablePaidRondelSlots(nationCurrentRondelSlot)) {
+        availableRondelSlots.add(
+          Action.rondel({
+            nation: nationName,
+            cost,
+            slot: paidRondelSlot.representation,
+          }),
+        );
+      }
+    }
+
     // Remove rondel positions that the player cannot afford.
     const { cash } = this.players[this.currentPlayerName];
-    for (const position of out) {
+    for (const position of availableRondelSlots) {
       if (position.payload.cost > cash) {
-        out.delete(position);
+        availableRondelSlots.delete(position);
       }
     }
-    // Remove factory if the nation cannot afford it.
-    const { treasury } = this.nations.get(nation);
-    let factoryAction = {};
-    for (const action of out) {
-      if (action.payload.slot === 'factory') {
-        factoryAction = action;
-      }
-    }
-    if (treasury < 5) {
-      out.delete(factoryAction);
-    }
-    return out;
+
+    return availableRondelSlots;
   }
 
-  extraRondelCost(index) {
-    if (this.baseGame === 'imperial') {
-      return index * 2;
-    }
+  costPerPaidRondelAction(nation) {
```

Comment:
In our new Onion representation of the code, should this function live elsewhere? Maybe on the `AvailableSlots` use case?

Reply:
[33b0e98](https://github.com/Thrillberg/imperial/pull/711/commits/33b0e98d928c58369c8fa89469759623b392478a)